### DISCUSSION TITLE
Add `line-casing-*` properties

### DISCUSSION THREAD

**Discussion by zstadler** [maplibre/maplibre#164] [Cross-project feature proposals]
Adding `line-casing-*` properties to lines, similar to the `circle-stroke-*` properties of circles would greatly simplify the styling of roads and highways. 

Currently, line casing for roads and highways requires a doubling the number of layers. The colored lines in the map below are defined by a set of layers, according to their class. The grey casing is defined by a second set of layers - one grey layer for each color layer.

![image](https://user-images.githubusercontent.com/1304610/208097563-2cbfd626-32f0-4401-b245-8baeb2fee79f.png)

In order to visualize bridges over bridges, like in the above map, a style is currently required to replicate a series of roads and highway layers for several values of the OSM `layer` tag. Such a style will further benefit from this addition.

#### Properties similar to `line-*` properties

- `line-casing-blur`
- `line-casing-cap`
- `line-casing-color`
- `line-casing-dasharray`
- `line-casing-gradient`
- `line-casing-join`
- `line-casing-miter-limit`
- `line-casing-offset` - positive numbers are to the outside
- `line-casing-opacity`
- `line-casing-pattern`
- `line-casing-round-limit`
- `line-casing-width`

Data-driven styling would be available according to their availability for the corresponding `line-*` property

#### Casing-specific property

| property | description | default |
|---|---|---|
| `line-casing-side` | One of `"none"`, `"left"`, `"right"`, `"both"` | `"both"`

Data-driven styling support
### Size
medium

**Comment by xabbu42** [comment]
This would be a great feature. Especially the point with the different layers is very important and makes current workarounds complex and closely tied to the data. For example you need to know in advance the exact levels you want to support. With this declarative style for casing you could just use the usual mechanisms of line-sort-key and presorted data to handle this in a generic way.

**Comment by zstadler** [comment]
An alternative approach to solve the problem could be to allow an array for the `paint` property. The renderer would paint them as if they were separate layers with the same layout definition.

For example
``` json
"paint": [
  {
    "line-color": "black",
    "line-width": {"base": 2, "stops": [[13, 6], [16, 24], [22, 96]]}
  },
  {
    "line-color": "#808080",
    "line-width": {
      "base": 2,
      "stops": [[10, 1], [12, 1.5], [13, 3], [16, 16], [22, 64]]
    }
  },
  {
    "line-color": "red",
    "line-width": {
      "base": 2,
      "stops": [[7, 0.5], [9, 1], [16, 8], [22, 32]]
    }
  }
]
```
Would produce an additional outer black casing for sections of the roads that actually have a `bridge` tag, compared to the image above.

![image](https://user-images.githubusercontent.com/1304610/208535183-ef8a921a-6ccf-4adf-b9be-c3c515930988.png)

**Reply by zekefarwell** [reply]
This seems like it could be a good idea because it would allow for multiple casings on given line.  It also might be applicable to more cartographic treatments than just casing since it would essentially be a generic sub-layers for paint properties.  I guess you wouldn't be able to apply layout properties though.  Not sure if that would be an issue.

**Reply by ZeLonewolf** [reply]
This is actually an excellent idea.  Thanks for suggesting it.  Is there an open issue in the main repo?

**Reply by MarcoLucafo-TomTom** [reply]
this would be great but it would be even better if we could define the rendering order of the sub-layers, so that e.g. a casing sub-layer is always drawn below the inline one (to solve the intersection problem). In conjunction with line-sort-key this would probably allow for a single-layer styling of an entire road network with proper z-order including bridges

**Comment by wipfli** [comment]
Thanks for this suggestion. I think it would be best to open an issue in MapLibre GL JS and Native if there is not already one for this issue.

**Comment by zstadler** [comment]
@wipfli
This is about an extension to the style spec. I'm not sure it should be turned directly into implementation-dependent issues.

**Reply by louwers** [reply]
The style spec is currently defined in the MapLibre GL JS repo. Whether this should be the case was a point of discussion during the last Technical Steering Comittee meeting. But for now I think the MapLibre GL JS repo is the right place to suggest additions to the style spec.

**Reply by louwers** [reply]
@zstadler The style spec has its own repo now!

https://github.com/maplibre/maplibre-gl-style-spec

And also good to know, there is a second round of voting: https://maplibre.org/news/2023-03-13-second-call-for-bounties/

**Comment by zstadler** [comment]
Sadly, I think this idea should be dropped.

As @ZeLonewolf described well in https://github.com/maplibre/maplibre-gl-js/issues/2108#issuecomment-1409278734, for a given OSM `layer` value, all casing should be drawn before any road lines are drawn. This allows same-`layer` roads of different colors to show a junction like this
<image width="30%" src="https://user-images.githubusercontent.com/1304610/217942220-9f84f965-9901-4d93-ba69-651d3b9eab94.png"/>
where the more important yellow road is drawn on top of the less important white road.

Unfortunately, this rendering is not possible with the idea I presented in this discussion. If it is applied for a simple junction like the above, the casing for the yellow road would make the junction look as if the white road goes under the yellow road:

<image width="30%" src="https://user-images.githubusercontent.com/1304610/217943258-5b211413-f2e2-4d6d-b869-f08386527419.png"/>

**Reply by ZeLonewolf** [reply]
Ugh, you're right, I forgot about intersections. Hence the need for cross-layer z-ordering.

**Reply by wipfli** [reply]
I did not follow exactly the proposal, but one cloud indicate a layer offset for casings. For example if you have 10 road types, use a casing offet of 10...

**Reply by xabbu42** [reply]
In the context of this proposal I would solve this by a line-casing-group-by expression.

**Reply by MarcoLucafo-TomTom** [reply]
I feel this might be circumvented by a similar extension of line-sort-key? Let's say you could assign different values to inline primitives with line-sort-key and to casing primitives with line-casing-sort-key. Then the renderer would sort them all and draw them based on the z-order we define.

**Comment by zstadler** [comment]
> I did not follow exactly the proposal, but one cloud indicate a layer offset for casings. For example if you have 10 road types, use a casing offet of 10...

I prefer a more structured approach, as I've described in https://github.com/maplibre/maplibre-gl-js/issues/2108#issuecomment-1410754642 and refined in https://github.com/maplibre/maplibre-gl-js/issues/2108#issuecomment-1413822379

**Comment by zstadler** [comment]
I'm dropping this proposal, as it does not address the problem it was supposed to address.
